pkg/diagnostics/server.go

```diff
@@ -41,25 +44,53 @@ func (s *diagnosticsServer) Start(stop <-chan struct{}) error {
 		resp.WriteHeader(http.StatusOK)
 	})
 	mux.Handle("/metrics", promhttp.InstrumentMetricHandler(s.metrics, promhttp.HandlerFor(s.metrics, promhttp.HandlerOpts{})))
-	if s.debugEndpoints {
+	if s.config.DebugEndpoints {
 		mux.HandleFunc("/debug/pprof/", pprof.Index)
 		mux.HandleFunc("/debug/pprof/cmdline", pprof.Cmdline)
 		mux.HandleFunc("/debug/pprof/profile", pprof.Profile)
 		mux.HandleFunc("/debug/pprof/symbol", pprof.Symbol)
 		mux.HandleFunc("/debug/pprof/trace", pprof.Trace)
 	}
 
-	httpServer := &http.Server{Addr: fmt.Sprintf(":%d", s.port), Handler: mux}
+	httpServer := &http.Server{Addr: fmt.Sprintf(":%d", s.config.ServerPort), Handler: mux}
 
-	diagnosticsServerLog.Info("starting diagnostic server", "interface", "0.0.0.0", "port", s.port)
+	if s.config.TlsEnabled {
+		cert, err := tls.LoadX509KeyPair(s.config.TlsCertFile, s.config.TlsKeyFile)
+		if err != nil {
+			return errors.Wrap(err, "failed to load TLS certificate")
+		}
+		tlsConfig := &tls.Config{Certificates: []tls.Certificate{cert}}
+		if tlsConfig.MinVersion, err = config_types.TLSVersion(s.config.TlsMinVersion); err != nil {
+			return err
+		}
+		if tlsConfig.MaxVersion, err = config_types.TLSVersion(s.config.TlsMaxVersion); err != nil {
+			return err
+		}
+		if tlsConfig.CipherSuites, err = config_types.TLSCiphers(s.config.TlsCipherSuites); err != nil {
+			return err
+		}
+		httpServer.TLSConfig = tlsConfig
+	}
+
+	diagnosticsServerLog.Info("starting diagnostic server", "interface", "0.0.0.0", "port", s.config.ServerPort, "tls", s.config.TlsEnabled)
 	errChan := make(chan error)
 	go func() {
 		defer close(errChan)
-		if err := httpServer.ListenAndServe(); err != nil {
-			if err.Error() != "http: Server closed" {
-				diagnosticsServerLog.Error(err, "terminated with an error")
-				errChan <- err
-				return
+		if s.config.TlsEnabled {
+			if err := httpServer.ListenAndServeTLS(s.config.TlsCertFile, s.config.TlsKeyFile); err != nil {
+				if err.Error() != "https: Server closed" {
+					diagnosticsServerLog.Error(err, "terminated with an error")
+					errChan <- err
+					return
+				}
+			}
+		} else {
+			if err := httpServer.ListenAndServe(); err != nil {
+				if err.Error() != "http: Server closed" {
+					diagnosticsServerLog.Error(err, "terminated with an error")
+					errChan <- err
+					return
+				}
```

Comment:
If any of you has a more elegant way to do this, I'm interested

Reply:
This should work no?

https://github.com/kumahq/kuma/blob/872f4c477c2fafe6954e6d6d5e2770e3e9fb9385/pkg/api-server/server.go#L344-L354

Reply:
I guess you could also extract the error handling out of the if as it's the same in both cases

Reply:
The documentation says that `ListenAndServeTLS` requires certfile and keyfile to exist, otherwise it will error .

See https://pkg.go.dev/net/http#ListenAndServeTLS 

Or am I missing something here ?

Reply:
The documentation says that `ListenAndServeTLS` requires certfile and keyfile to exist, otherwise it will error .

See https://pkg.go.dev/net/http#ListenAndServeTLS 

Or am I missing something here ?

Reply:
I meant doing:

```go
var err error
if tlsEnabled {
    err = httpServer.ListenAndServeTLS(s.config.TlsCertFile, s.config.TlsKeyFile)
} else {
    err = httpServer.ListenAndServe()
}
if err != nil {
 		switch err { 
 		case http.ErrServerClosed: 
 			log.Info("shutting down server") 
 		default: 
 			log.Error(err, "could not start diagnostic server") 
 			errChan <- err 
 		} 
}
```